drivers/entropy/entropy_cc13xx_cc26xx.c

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2019 Brett Witherspoon
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <device.h>
+#include <entropy.h>
+#include <irq.h>
+#include <ring_buffer.h>
+#include <sys_io.h>
+
+#include <driverlib/prcm.h>
+#include <driverlib/trng.h>
+
+struct entropy_cc13xx_cc26xx_data {
+	struct k_sem lock;
+	struct k_sem sync;
+	struct ring_buf pool;
+	u8_t data[CONFIG_ENTROPY_CC13XX_CC26XX_POOL_SIZE];
+};
+
+DEVICE_DECLARE(entropy_cc13xx_cc26xx);
+
+static inline struct entropy_cc13xx_cc26xx_data *
+get_dev_data(struct device *dev)
+{
+	return dev->driver_data;
+}
+
+static int entropy_cc13xx_cc26xx_get_entropy(struct device *dev, u8_t *buf,
+					     u16_t len)
+{
+	struct entropy_cc13xx_cc26xx_data *data = get_dev_data(dev);
+	u32_t cnt;
+
+	TRNGIntEnable(TRNG_NUMBER_READY);
+
+	while (len) {
+		k_sem_take(&data->lock, K_FOREVER);
+		cnt = ring_buf_get(&data->pool, buf, len);
+		k_sem_give(&data->lock);
+
+		if (cnt) {
+			buf += cnt;
+			len -= cnt;
+		} else {
+			k_sem_take(&data->sync, K_FOREVER);
+		}
+	}
+
+	return 0;
+}
+
+static void entropy_cc13xx_cc26xx_isr(void *arg)
+{
+	struct entropy_cc13xx_cc26xx_data *data = get_dev_data(arg);
+	u32_t src, cnt, off;
+	u32_t num[2];
+
+	/* Interrupt service routine as described in TRM section 18.16.1.3.2 */
+	src = TRNGStatusGet();
+
+	if (src & TRNG_NUMBER_READY) {
+		num[1] = TRNGNumberGet(TRNG_HI_WORD);
+		num[0] = TRNGNumberGet(TRNG_LOW_WORD);
+
+		cnt = ring_buf_put(&data->pool, (u8_t *)num, sizeof(num));
+
+		/* When pool is full disable interrupt and stop reading numbers */
+		if (cnt != sizeof(num)) {
+			TRNGIntDisable(TRNG_NUMBER_READY);
+		}
+
+		k_sem_give(&data->sync);
+	}
+
+	if (src & TRNG_FRO_SHUTDOWN) {
+		/* Disabled FROs */
+		off = sys_read32(TRNG_BASE + TRNG_O_ALARMSTOP);
+		/* Clear alarms */
+		sys_write32(0, TRNG_BASE + TRNG_O_ALARMMASK);
+		sys_write32(0, TRNG_BASE + TRNG_O_ALARMSTOP);
+		/* De-tune FROs */
+		sys_write32(off, TRNG_BASE + TRNG_O_FRODETUNE);
+		/* Re-enable FROs */
+		sys_write32(off, TRNG_BASE + TRNG_O_FROEN);
+		/* Clear shutdown */
+		TRNGIntClear(TRNG_FRO_SHUTDOWN);
```

Comment:
The documentation for TRNGIntClear() in driverlib/trng.h suggests to clear the interrupt early on in the ISR to allow the event clear to propagate to the NVIC before returning from the ISR, so it may be better to move line 90 to somewhere before line 81.

Reply:
This is a good point. I am attempting to duplicate Figure 18-3 on page 1402 of the TRM which has it after re-enabling the FROs which also right before the return. I will investigate if this order is important, but it does seem better to do as you suggested.

Reply:
After reviewing section 16.4, it seems the earliest I should clear the SHUTDOWN_OVF event would be after clearing the bits in the ALARMSTOP register. My reasoning is that IRQFLAGSTAT.SHUTDOWN_OVF is triggered when ALARMCNT.SHUTDOWN_CNT is greater then ALARMCNT.SHUTDOWN_THR and ALARMCNT.SHUTDOWN_CNT is the number of bits set in the ALARMSTOP register. I am not sure if the interrupt would be triggered again if it was cleared before the ALARMCNT register, but if it was it would just repeat the process with zero FRO's being disabled. Unfortunately, it's not really possible for me to test this code path.

Reply:
TRNGCC26X2_basicHwiFxn() in source/ti/drivers/trng/TRNGCC26X2.c from the TI SDK does the opposite and clears the interrupt first. I just thought it may be best to follow the same approach If you are unable to verify this code path. The way it is currently the TRNGIntClear() call is too close to the end of the ISR for comfort.

Reply:
Awesome. I will do the same.